controllers/topology/desired_state.go

```diff
@@ -28,143 +28,164 @@ import (
 // NOTE: We are assuming all the required objects are provided as input; also, in case of any error,
 // the entire compute operation operation will fail. This might be improved in the future if support for reconciling
 // subset of a topology will be implemented.
-func (r *ClusterReconciler) computeDesiredState(_ context.Context, class *clusterTopologyClass, current *clusterTopologyState) (*clusterTopologyState, error) {
+func (r *ClusterReconciler) computeDesiredState(ctx context.Context, s *scope.Scope) (*scope.ClusterState, error) {
 	var err error
-	desiredState := &clusterTopologyState{}
+	desiredState := &scope.ClusterState{
+		ControlPlane: &scope.ControlPlaneState{},
+	}
 
 	// Compute the desired state of the InfrastructureCluster object.
-	if desiredState.infrastructureCluster, err = computeInfrastructureCluster(class, current); err != nil {
+	if desiredState.InfrastructureCluster, err = computeInfrastructureCluster(ctx, s); err != nil {
 		return nil, err
 	}
 
-	// If the ControlPlane object requires it, compute the InfrastructureMachineTemplate for the ControlPlane.
-	if class.clusterClass.Spec.ControlPlane.MachineInfrastructure != nil {
-		if desiredState.controlPlane.infrastructureMachineTemplate, err = computeControlPlaneInfrastructureMachineTemplate(class, current); err != nil {
+	// If the clusterClass mandates the controlPlane has infrastructureMachines, compute the InfrastructureMachineTemplate for the ControlPlane.
+	if s.Blueprint.HasControlPlaneInfrastructureMachine() {
+		if desiredState.ControlPlane.InfrastructureMachineTemplate, err = computeControlPlaneInfrastructureMachineTemplate(ctx, s); err != nil {
 			return nil, err
 		}
 	}
 
 	// Compute the desired state of the ControlPlane object, eventually adding a reference to the
 	// InfrastructureMachineTemplate generated by the previous step.
-	if desiredState.controlPlane.object, err = computeControlPlane(class, current, desiredState.controlPlane.infrastructureMachineTemplate); err != nil {
+	if desiredState.ControlPlane.Object, err = computeControlPlane(ctx, s, desiredState.ControlPlane.InfrastructureMachineTemplate); err != nil {
 		return nil, err
 	}
 
 	// Compute the desired state for the Cluster object adding a reference to the
 	// InfrastructureCluster and the ControlPlane objects generated by the previous step.
-	desiredState.cluster = computeCluster(current, desiredState.infrastructureCluster, desiredState.controlPlane.object)
+	desiredState.Cluster = computeCluster(ctx, s, desiredState.InfrastructureCluster, desiredState.ControlPlane.Object)
 
-	// Compute the desired state of the MachineDeployment objects for the worker nodes.
-	if current.cluster.Spec.Topology.Workers == nil || len(current.cluster.Spec.Topology.Workers.MachineDeployments) == 0 {
+	// If required by the blueprint, compute the desired state of the MachineDeployment objects for the worker nodes.
+	if s.Blueprint.HasMachineDeployments() {
```

Comment:
That also seems to be inverted (like the implementation). So I guess we're fine :)

Reply:
fixed in https://github.com/kubernetes-sigs/cluster-api/pull/5128